mypy/build.py

```diff
@@ -778,95 +785,188 @@ def is_file(path: str) -> bool:
+        if not is_file(path):
             return False
-    return True
+
+        return True
+
+    def _verify_namespace(self, path: str) -> bool:
+        if os.path.isdir(path):
+            files = set(list_dir(path) or [])  # type: Set[str]
+            if not ('__init__.py' in files or '__init__.pyi' in files):
+                return True
+
+        return False
+
+
+def is_pkg_path(path: str) -> bool:
+    return path.endswith(('__init__.py', '__init__.pyi'))
+
+
+def is_module_path(path: str) -> bool:
+    return not is_pkg_path(path) and path.endswith(('.py', '.pyi'))
+
+
+def is_namespace_path(path: str) -> bool:
+    return path.endswith(os.path.sep)
+
+
+class ModuleDiscovery:
+    def __init__(self,
+                 lib_path: Iterable[str],
+                 namespaces_allowed: bool = False) -> None:
+
+        self.lib_path = tuple(os.path.normpath(p) for p in lib_path)  # type: Tuple[str, ...]
+        self.namespaces_allowed = namespaces_allowed
+        self._find_module_cache = {}  # type: Dict[str, List[str]]
+
+    def find_module(self, id: str) -> Optional[str]:
+        paths = self._find_module(id)
+        if paths:
+            return paths[0]
+        else:
+            return None
+
+    def find_modules_recursive(self, module: str) -> List[BuildSource]:
+        """
+        Discover module and all it's children
+        Remove duplicates from discovered paths
+        """
+        hits = set()  # type: Set[str]
+        result = []  # type: List[BuildSource]
+        for src in self._find_modules_recursive(module):
+            if src.module not in hits:
+                hits.add(src.module)
+                result.append(src)
+        return result
+
+    def _find_modules_recursive(self, module: str) -> List[BuildSource]:
+        module_paths = self._find_module(module)
+
+        srcs = []  # type: List[BuildSource]
+        for path in module_paths:
+            if is_module_path(path) or is_pkg_path(path):
+                srcs.append(BuildSource(path, module, None))
+
+            if is_pkg_path(path):
+                path = dirname(path)
+                for submodule in self._find_submodules(module, path):
+                    srcs += self._find_modules_recursive(submodule)
+            elif is_namespace_path(path):
+                for submodule in self._find_submodules(module, path):
+                    srcs += self._find_modules_recursive(submodule)
+
+        return srcs
+
+    def _find_submodules(self, module: str, path: str) -> Iterator[str]:
+        for item in list_dir(path) or []:
+            if item == '__init__.py' or item == '__init__.pyi':
+                continue
+
+            if item.endswith(tuple(PYTHON_EXTENSIONS)):
+                item = item.split('.')[0]
+
+            yield module + '.' + item
+
+    def _collect_paths(self, paths: List[str], last_comp: str) -> List[str]:
+        """
+        Collect all available module paths
```

Comment:
"Return the path (or namespace paths) for given (sub)module name, relative to given parent paths."